git/odb/tree.go

```diff
@@ -0,0 +1,123 @@
+package odb
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"syscall"
+)
+
+// Tree encapsulates a Git tree object.
+type Tree struct {
+	// Entries is the list of entries held by this tree.
+	Entries []*TreeEntry
+}
+
+var _ Object = (*Tree)(nil)
+
+// Type implements Object.ObjectType by returning the correct object type for
+// Trees, TreeObjectType.
+func (t *Tree) Type() ObjectType { return TreeObjectType }
+
+// Decode implements Object.Decode and decodes the uncompressed tree being
+// read. It returns the number of uncompressed bytes being consumed off of the
+// stream, which should be strictly equal to the size given.
+//
+// If any error was encountered along the way, that will be returned, along with
+// the number of bytes read up to that point.
+func (t *Tree) Decode(from io.Reader, size int64) (n int, err error) {
+	buf := bufio.NewReader(from)
+
+	var entries []*TreeEntry
+	for {
+		modes, err := buf.ReadString(' ')
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			return n, err
+		}
+		n += len(modes)
+		modes = strings.TrimSuffix(modes, " ")
+		modes = fmt.Sprintf("%06s", modes)
+
+		mode, _ := strconv.ParseInt(modes, 8, 32)
+
+		fname, err := buf.ReadString('\x00')
+		if err != nil {
+			return n, err
+		}
+		n += len(fname)
+		fname = strings.TrimSuffix(fname, "\x00")
+
+		var sha [20]byte
+		if _, err = buf.Read(sha[:]); err != nil {
+			return n, err
+		}
+		n += 20
+
+		var typ ObjectType
+		switch mode & syscall.S_IFMT {
+		case syscall.S_IFREG:
+			typ = BlobObjectType
+		case syscall.S_IFDIR:
+			typ = TreeObjectType
+		case syscall.S_IFLNK:
+			typ = BlobObjectType
+		default:
+			panic(fmt.Sprintf("git/odb: unknown object type: %q %q", modes, fname))
+		}
```

Comment:
> and found that the representation is commit:

Sort of. Gitlink entries do point to commits, but note that this is the one object pointer in Git that doesn't require that the object be present (because it's probably in a submodule). So you wouldn't want to recurse into it the same way as the other tree entries.